File: backend/src/main/java/com/paymentplatform/payment/service/PaymentService.java
package com.paymentplatform.payment.service;

import com.paymentplatform.payment.dto.PaymentRequest;
import com.paymentplatform.payment.dto.PaymentResponse;
import com.paymentplatform.payment.enums.PaymentMethodType;
import com.paymentplatform.payment.enums.PaymentProvider;
import com.paymentplatform.payment.enums.PaymentStatus;
import com.paymentplatform.payment.exec.BankTransferExec;
import com.paymentplatform.payment.exec.CardPaymentExec;
import com.paymentplatform.payment.exec.MobileMoneyPaymentExec;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
@Slf4j
public class PaymentService {

    private final CardPaymentExec cardPaymentExec;
    private final MobileMoneyPaymentExec mobileMoneyPaymentExec;
    private final BankTransferExec bankTransferExec;

    public PaymentResponse processPayment(PaymentRequest request) {
        try {
            if (request.getPaymentMethod() == PaymentMethodType.CARD) {
                PaymentResponse cardResponse = cardPaymentExec.execute(request);

                if (cardResponse.isSuccess()) {
                    return cardResponse;
                }

                return tryFallback(request, cardResponse);
            }
            if (request.getPaymentMethod() == PaymentMethodType.MOBILE_MONEY) {
                return mobileMoneyPaymentExec.execute(request);
            }
            if (request.getPaymentMethod() == PaymentMethodType.BANK_TRANSFER) {
                return bankTransferExec.execute(request);
            }
            throw new IllegalArgumentException("Type de paiement non supporté: " + request.getPaymentMethod());
        } catch (Exception ex) {
            log.error("Payment processing failed", ex);
            return PaymentResponse.builder()
                    .paymentId(java.util.UUID.randomUUID())
                    .status(PaymentStatus.FAILED)
                    .transactionReference("ERR-" + java.util.UUID.randomUUID().toString().substring(0, 8).toUpperCase())
                    .externalTransactionId(null)
                    .amount(request != null ? request.getAmount() : 0)
                    .currency(request != null ? request.getCurrency() : "MGA")
                    .paymentMethod(request != null && request.getPaymentMethod() != null ? request.getPaymentMethod() : com.paymentplatform.payment.enums.PaymentMethodType.CARD)
                    .provider(null)
                    .message("Une erreur interne est survenue lors du paiement. Veuillez réessayer.")
                    .success(false)
                    .token(request != null ? request.getToken() : null)
                    .build();
        }
    }

    private PaymentResponse tryFallback(PaymentRequest request, PaymentResponse original) {
        log.warn("Card payment declined ({}), attempting cross-channel fallback", original.getMessage());

        if (request.getMobileMoneyPhone() != null && !request.getMobileMoneyPhone().isBlank()) {
            log.info("Fallback: trying mobile money via {}", request.getMobileMoneyOperator());
            PaymentRequest mmRequest = PaymentRequest.builder()
                    .token(request.getToken())
                    .merchantApiKey(request.getMerchantApiKey())
                    .paymentMethod(PaymentMethodType.MOBILE_MONEY)
                    .provider(request.getProvider())
                    .amount(request.getAmount())
                    .currency(request.getCurrency())
                    .mobileMoneyPhone(request.getMobileMoneyPhone())
                    .mobileMoneyOperator(request.getMobileMoneyOperator())
                    .description("Fallback depuis paiement par carte: " + original.getTransactionReference())
                    .isOneClick(request.getIsOneClick())
                    .build();

            PaymentResponse mmResponse = mobileMoneyPaymentExec.execute(mmRequest);
            if (mmResponse.isSuccess()) {
                 mmResponse.setFallbackUsed(true);
                mmResponse.setFallbackFromProvider(original.getProvider() != null ? original.getProvider().name() : "VISA");
                mmResponse.setFallbackToProvider(mmResponse.getProvider() != null ? mmResponse.getProvider().name() : "UNKNOWN");
                log.info("Fallback succeeded: mobile money payment accepted");
                return mmResponse;
            }
        }

        if (request.getDestinationAccount() != null && !request.getDestinationAccount().isBlank()) {
            log.info("Fallback: trying bank transfer to {}", request.getDestinationAccount());
            PaymentRequest btRequest = PaymentRequest.builder()
                    .token(request.getToken())
                    .merchantApiKey(request.getMerchantApiKey())
                    .paymentMethod(PaymentMethodType.BANK_TRANSFER)
                    .provider(request.getProvider())
                    .amount(request.getAmount())
                    .currency(request.getCurrency())
                    .destinationAccount(request.getDestinationAccount())
                    .description("Fallback depuis paiement par carte: " + original.getTransactionReference())
                    .build();

            PaymentResponse btResponse = bankTransferExec.execute(btRequest);
            if (btResponse.isSuccess()) {
                btResponse.setFallbackUsed(true);
                btResponse.setFallbackFromProvider(original.getProvider() != null ? original.getProvider().name() : "VISA");
                btResponse.setFallbackToProvider(btResponse.getProvider() != null ? btResponse.getProvider().name() : "UNKNOWN");
                log.info("Fallback succeeded: bank transfer accepted");
                return btResponse;
            }
        }

        log.warn("All fallback channels failed, returning original card decline response");
        return original;
    }
}

File: frontend/src/services/paymentService.ts
import api from './api';
import type { TokenizeResponse } from './vaultService';

export type PaymentMethodType = 'CARD' | 'MOBILE_MONEY' | 'BANK_TRANSFER';
export type PaymentProvider = 'STRIPE' | 'ADYEN' | 'ORANGE_MONEY' | 'MTN_MOBILE_MONEY' | 'MPESA' | 'VISA' | 'MASTERCARD';
export type PaymentStatus = 'PENDING' | 'SUCCESS' | 'FAILED' | 'CANCELLED' | 'REFUNDED';

export interface MobileMoneyOperator {
    name: string;
    displayName: string;
    countryCallingCode: string;
}

export interface PaymentRequest {
    token: string;
    merchantApiKey?: string;
    paymentMethod: PaymentMethodType;
    provider?: PaymentProvider;
    amount: number;
    currency: string;
    customerId?: string;
    mobileMoneyPhone?: string;
    mobileMoneyOperator?: string;
    cardHolderName?: string;
    destinationAccount?: string;
    idempotencyKey?: string;
    description?: string;
    threeDsAuthCode?: string;
    previousTransactionId?: string;
    isOneClick?: boolean;
}

export interface PaymentResponse {
    paymentId: string;
    status: PaymentStatus;
    transactionReference: string;
    amount: number;
    currency: string;
    paymentMethod: PaymentMethodType;
    provider?: PaymentProvider;
    message: string;
    success: boolean;
    token?: string;
    externalTransactionId?: string;
    riskScore?: number;
    threeDsRequired?: boolean;
    riskLevel?: string;
    fraudRecommendation?: string;
}

export interface PaymentProvidersResponse {
    cardProviders: PaymentProvider[];
    methodTypes: PaymentMethodType[];
}

export const PROVIDER_LABELS: Record<string, string> = {
    STRIPE: 'Stripe',
    ADYEN: 'Adyen',
    ORANGE_MONEY: 'Orange Money',
    MTN_MOBILE_MONEY: 'MTN Mobile Money',
    MPESA: 'M-Pesa',
    VISA: 'Visa',
    MASTERCARD: 'Mastercard',
};

const paymentService = {
    processPayment: async (request: PaymentRequest): Promise<PaymentResponse> => {
        const idempotencyKey = request.idempotencyKey || `pay-${Date.now()}-${Math.random().toString(36).substring(2, 10)}`;
        const response = await api.post('/payments/process', request, {
            headers: { 'Idempotency-Key': idempotencyKey },
        });
        return response.data;
    },
    getPaymentProviders: async (): Promise<PaymentProvidersResponse> => {
        const response = await api.get('/payments/providers');
        return response.data;
    },
    tokenizeCard: async (cardData: {
        pan: string;
        expiryMonth: string;
        expiryYear: string;
        cardHolderName: string;
    }): Promise<TokenizeResponse> => {
        const response = await api.post('/vault/tokenize', cardData);
        return response.data;
    },
};

export default paymentService;
